windows/src/developer/TIKE/compile/CompileKeymanWeb.pas

```diff
@@ -2059,5 +2152,132 @@ function GetKeymanWebCompiledFileName(const FileName: WideString): WideString;
+    Result := '';
+end;
+
+///
+/// Converts a modifier bit mask integer into its component bit flags
+///
+/// @param FBitMask A KMX modifier bitmask value
+///
+/// @return string of JavaScript code, e.g. 'modCodes.SHIFT | modCodes.CTRL /* 0x0030 */'
+///
+function TCompileKeymanWeb.FormatModifierAsBitflags(FBitMask: Cardinal): string;
+const
+  mask: array[0..14] of string = (
+    'LCTRL',             // 0X0001
+    'RCTRL',             // 0X0002
+    'LALT',              // 0X0004
+    'RALT',              // 0X0008
+
+    'SHIFT',             // 0X0010
+    'CTRL',              // 0X0020
+    'ALT',               // 0X0040
+
+    '???',               // Reserved
+
+    'CAPS',              // 0X0100
+    'NO_CAPS',           // 0X0200
+
+    'NUM_LOCK',          // 0X0400
+    'NO_NUM_LOCK',       // 0X0800
+
+    'SCROLL_LOCK',       // 0X1000
+    'NO_SCROLL_LOCK',    // 0X2000
+
+    'VIRTUAL_KEY'        // 0X4000
+  );
+var
+  i: Integer;
+begin
+  //TODO: We need to think about mnemonic layouts which are incompletely supported at present
+  //tavultesoft.keymanweb.osk.
+
+  Result := '';
+
+  for i := 0 to High(mask) do
+  begin
+    if FBitMask and (1 shl i) <> 0 then
+    begin
+      if Result <> '' then Result := Result + ' | ';
+      Result := Result + 'modCodes.'+mask[i];
+    end;
+  end;
+
+  if Result = '' then
+    Result := '0';
+
+  Result := Result + ' /* 0x' + IntToHex(FBitMask, 4) + ' */';
+end;
+
+///
+/// Converts a key value into a constant
+///
+/// @param key A virtual key code
+///
+/// @return string of JavaScript code, e.g. 'keyCodes.K_A /* 0x41 */'
+///
+function TCompileKeymanWeb.FormatKeyAsString(key: Integer): string;
+begin
+  if (key <= 255) and (KMWVKeyNames[key] <> '')
+    then Result := 'keyCodes.'+KMWVKeyNames[key]+ ' /* 0x' + IntToHex(key, 2) + ' */'
+    else Result := IntToHex(key, 2);
+end;
+
+///
+/// Determine the modifiers used in the target keyboard and return a bitmask
+/// representing them, or an integer value when not in debug mode
+///
+/// @return string of JavaScript code, e.g. 'modCodes.SHIFT | modCodes.CTRL /* 0x0030 */'
+///
+function TCompileKeymanWeb.GetKeyboardModifierBitmask: string;
+var
+  i: Integer;
+  gp: PFILE_GROUP;
+  j: Integer;
+  kp: PFILE_KEY;
+  FBitMask: Integer;
+begin
+  FBitMask := 0;
+  gp := fk.dpGroupArray;
+  for i := 0 to fk.cxGroupArray-1 do
+  begin
+    if gp.fUsingKeys then
+    begin
+      kp := gp.dpKeyArray;
+      for j := 0 to gp.cxKeyArray-1 do
+      begin
+        if not RuleIsExcludedByPlatform(kp) then
+          FBitMask := FBitMask or JavaScript_Shift(kp, fMnemonic);
+        Inc(kp);
+      end;
+    end;
+    Inc(gp);
+  end;
+
+  if ((FBitMask and KMX_MASK_MODIFIER_CHIRAL) <> 0) and
+    ((FBitMask and KMX_MASK_MODIFIER_NONCHIRAL) <> 0) then
+  begin
+    ReportError(0, CWARN_DontMixChiralAndNonChiralModifiers, 'This keyboard contains Ctrl,Alt and LCtrl,LAlt,RCtrl,RAlt sets of modifiers. Use only one or the other set for web target.');
```

Comment:
@darcywong00 This line here warns on use of both chiral and non-chiral modifiers.

Reply:
That checks for combinations like CTRL + left CTRL, correct?

I was more wondering about blocking developers from making a keyboard with something like
left alt + right ctrl + ...

Reply:
Okay, just added code to test that into the kmx compiler (because we want that test in both Keyman Desktop keyboards and KeymanWeb keyboards).